flax/nn/utils.py

```diff
@@ -95,3 +96,120 @@ def _level_of_value(xs):
       level = _trace_level(x._trace.master)
       max_level = max(level, max_level)
   return max_level
+
+
+def flatten_dict(input_dict: Dict[str, Any], prefix: str = "") -> Dict[str, Any]:
+  """Flattens the keys of a nested dictionary."""
+  output_dict = {}
+  for key, value in input_dict.items():
+    nested_key = "{}/{}".format(prefix, key) if prefix else key
+    if isinstance(value, dict):
+      output_dict.update(flatten_dict(value, prefix=nested_key))
+    else:
+      output_dict[nested_key] = value
+  return output_dict
+
+
+def _name_idx(name: str):
+  """Returns the layer index of the parameter name."""
+  index = name[name.find('_') + 1 : name.find('/')]
```

Comment:
cc @levskaya @jheek weird stuff that happens when we autoincrement layer names as strings -- "everything becomes part of the API"

Reply:
I think this is bad style. New layers might not follow this convention and the param info shouldn't rely on it. I actually have good experience with just sorting alphabetical by name.

Reply:
@rolandgvc - I wouldn't try to sort parameters by the "layer order", in general the "order" of layers is a partial order given that arbitrary dataflow can happen in complex modules.  Also this ordering information is lost for manually named layers.

@avital - the alternative to names is an opaque tree-structure-based serialization (which we used in trax) and it is a complete nightmare to work with when debugging.  Pragmatically, I am happy to suffer hyrum's law if it enables vastly more pleasant debugging.  I want a human-navigable "at rest" representation of models.

Reply:
I believe we can get "the best of both worlds" with easy
introspection while allowing people to look through ordered lists when
those are semantically meaningful.

In the meanwhile I propose we add a strong TODO comment here saying that 
we should use this as a use-case for the ongoing API rewrite considerations.

Reply:
This could also be resolved by having the params as OderedDicts instead so they can keep the order of the applied modules by default.

Reply:
Let's just add a comment for now and I think we can merge this.